mm2src/common/executor/abortable_system/abortable_queue.rs

```diff
@@ -153,32 +160,28 @@ impl QueueInner {
     /// Inserts the given future `handle`.
     fn insert_handle(&mut self, handle: oneshot::Sender<()>) -> FutureId {
         match self.finished_futures.pop() {
-            Some(finished_id) => {
+            // We can reuse the given `finished_id`.
+            Ok(finished_id) => {
                 self.abort_handlers[finished_id] = handle;
```

Comment:
what about doing a check here before indexing directly ?

Reply:
Good note! If there is a wrong `FutureId` in `finished_futures`, we can get 2 problems:
1) If `finished_id < abort_handlers.len()`, we'll reset a valid future handle at `abort_handlers[finished_id]`, and it will be aborted - that can lead to a very complicated debugging process;
2) If `finished_id >= abort_handlers.len()`, we'll get a panic.

If the `2)` option happens, it means that we are in a wrong state, and later some other wrong `finished_id` can appear and lead to the `1)` option. I think it's worth to panic in order to let the developers to quick fix it.

Reply:
Alright then. Thanks

Reply:
It's still an open topic, so if there is a way to improve it, I'll be glad to read 🙂

Reply:
right, I think we can handle the case for `2)` I mean we should do `nothing` if option `2)` is true and log something like `current index doesn't belong to any future` instead of panic. WDYT?

Reply:
That's interesting! Will consider implementing it tomorrow :) Thank you!

Reply:
I've added the checking, thank you for the note!